tensorflow/examples/android/src/org/tensorflow/demo/env/ImageUtils.java

```diff
@@ -98,6 +98,60 @@ public static void saveBitmap(final Bitmap bitmap, final String filename) {
   // Always prefer the native implementation if available.
   private static boolean useNativeConversion = true;
 
+  public static void convertYUV420SPToARGB8888(
+      byte[] input,
+      int[] output,
+      int width,
+      int height) {
+
+    if (useNativeConversion) {
+      try {
+        ImageUtils.convertYUV420SPToARGB8888(input, output, width, height, false);
+        return;
+      } catch (UnsatisfiedLinkError e) {
+        LOGGER.w("Native YUV420SP -> RGB implementation not found, falling back to Java implementation");
+        useNativeConversion = false;
+      }
+    }
+
+    // Java implementation of YUV420SP to ARGB8888 converting
+    final int frameSize = width * height;
+    for (int j = 0, yp = 0; j < height; j++) {
+      int uvp = frameSize + (j >> 1) * width, u = 0, v = 0;
+      for (int i = 0; i < width; i++, yp++) {
+        int y = (0xff & ((int) input[yp])) - 16;
+        if (y < 0)
+          y = 0;
+        if ((i & 1) == 0) {
+          v = (0xff & input[uvp++]) - 128;
+          u = (0xff & input[uvp++]) - 128;
+        }
+
+        int y1192 = 1192 * y;
```

Comment:
It seems like everything from here to the bottom of the loop can be shared between both conversion functions in a common helper method. The only difference between the two conversion types should be how it initializes the y,u,v variables per output pixel.